controller/root.go

```diff
@@ -0,0 +1,152 @@
+	if c.endpoints == nil {
+		roots, err := getRoot(ctx, c.FileHandler)
+		if err != nil || roots == nil {
+			log.Error(ctx, map[string]interface{}{
+				"err":  errs.WithStack(err),
+				"file": embeddedSwaggerSpecFile,
+			}, err.Error())
+			return jsonapi.JSONErrorResponse(ctx, errs.WithStack(err))
+		}
+		c.endpoints = roots
+	}
+	return ctx.OK(&app.RootSingle{Data: c.endpoints})
+}
+
+// Get a list of all endpoints formatted to json api format.
+func getRoot(ctx *app.ListRootContext, fileHandler asseter) (*app.Root, error) {
+	// Get an unmarshal swagger specification
+	swaggerJSON, err := fileHandler.Asset(embeddedSwaggerSpecFile)
+	if err != nil {
+		// TODO(tinakurian): fix error handling
+		return nil, errors.NewNotFoundError("file", embeddedSwaggerSpecFile)
+	}
+
+	var result map[string]interface{}
+	err = json.Unmarshal([]byte(swaggerJSON), &result)
+	if err != nil {
+		return nil, errs.Wrapf(err, "unable to unmarshal the file with id "+"'"+embeddedSwaggerSpecFile+"'")
+	}
+
+	// Get and iterate over paths from swagger specification.
+	swaggerPaths, ok := result["paths"]
+	if !ok {
+		return nil, errors.NewInternalErrorFromString("field `paths` could be found in swagger specification")
+	}
+
+	swaggerPathz, ok := swaggerPaths.(map[string]interface{})
+	if !ok {
+		return nil, errors.NewInternalErrorFromString("unable to assert concrete type map for field `paths` in swagger specifiation")
+	}
+
+	// the path map stores paths as key and URLs as values
+	namedPaths := make(map[string]interface{})
+	for path, swaggerPath := range swaggerPathz {
+
+		// Currently not supporting endpoints that contain parameters.
+		if !strings.Contains(path, "{") {
+			key := path
+
+			pathsObj, ok := swaggerPath.(map[string]interface{})
+			if !ok {
+				return nil, errors.NewInternalErrorFromString("unable to assert concrete type map for field `paths` in swagger specifiation")
+			}
+
+			// Get the x-tag value. If the tag exists, use it as path name.
+			xtagObj, ok := pathsObj["x-tag"]
+			if ok {
+				xtag, ok := xtagObj.(string)
+				if !ok {
+					return nil, errors.NewInternalErrorFromString("unable to assert concrete type string for field `x-tag` in swagger specifiation")
+				}
+
+				key = xtag
+			}
+
+			// cleanup the key to conform to JSONAPI member names
+			key = jsonapi.FormatMemberName(key)
+
+			// Set the related field and link objects for each name.
+			namedPaths[key] = map[string]interface{}{
+				"links": map[string]string{
+					"related": path,
+				},
+			}
+		}
+	}
+
+	basePathObj, ok := result["basePath"]
+	if !ok {
+		return nil, errors.NewInternalErrorFromString("field `basePath` could be found in swagger specification")
+	}
+
+	basePath, ok := basePathObj.(string)
+	if !ok {
+		return nil, errors.NewInternalErrorFromString("unable to assert concrete type string for field `basePath` in swagger specifiation")
+	}
+
+	return &app.Root{
+		Type:          "endpoints",
+		ID:            uuid.NewV4(),
```

Comment:
Why do we have an ID field?

Reply:
Ugly, isn't it? But it is required to conform to JSONAPI. http://jsonapi.org/format/#document-resource-objects

Reply:
That's sad :/